libqrexec/exec.c

```diff
@@ -105,4 +139,240 @@ void do_fork_exec(const char *cmdline, int *pid, int *stdin_fd, int *stdout_fd,
+        socket_len = (socklen_t)(offsetof(struct sockaddr_un, sun_path) + sizeof buf);
+    } else {
+        memcpy(remote.sun_path, connect_path, total_path_length);
+        remote.sun_path[total_path_length] = '\0';
+        socket_len = (socklen_t)(offsetof(struct sockaddr_un, sun_path) + total_path_length + 1);
+    }
+
+    do
+       result = connect(s, (struct sockaddr *) &remote, socket_len);
+    while (result < 0 && (errno == EINTR || errno == EWOULDBLOCK || errno == EAGAIN));
+    dummy_errno = errno;
+    unlink(buf);
+out:
+    buf[path_separator_offset] = '\0';
+    rmdir(buf);
+    errno = dummy_errno;
+    return result;
+}
+
+/* A parsed, mostly-validated RPC command. */
+struct qrexec_parsed_command {
+    /* NULL if and only if we are the fork server.  Otherwise, a NUL-terminated string. */
+    const char *const username;
+    /* Command line.  Never NULL.  NUL-terminated. Does not include "QUBESRPC ". */
+    char *const command;
+    /* Service descriptor.  Identical to `command` unless the "nogui:" prefix is present, in which case it points
+     * after the colon.  Always points to the start of the service name. */
+    const char *service_descriptor;
+    /* Remote domain.  NUL-terminated.  Always equal to `service_descriptor + service_descriptor_length + 1`. */
+    const char *const remote_domain;
+    /* Size of service_descriptor (the service name + argument).  Guaranteed to be <= MAX_SERVICE_NAME_LEN. */
+    size_t const service_descriptor_length;
+};
+
+static int execute_parsed_qubes_rpc_command(const struct qrexec_parsed_command *const command, int *const pid, int *const stdin_fd, int *const stdout_fd, int *const stderr_fd, struct buffer *stdin_buffer);
+
+static const char *skip_nogui(const char *cmdline) {
+    return strncmp(cmdline, NOGUI_CMD_PREFIX, NOGUI_CMD_PREFIX_LEN) ? cmdline : cmdline + NOGUI_CMD_PREFIX_LEN;
+}
+
+int execute_qubes_rpc_command(char *cmdline, int *pid, int *stdin_fd,
+        int *stdout_fd, int *stderr_fd, bool strip_username, struct buffer *stdin_buffer) {
+    const char *service_descriptor, *remote_domain;
+    char *realcmd;
+    size_t service_descriptor_length;
+    {
+        char *end_user;
+        if (strip_username) {
+            end_user = strchr(cmdline, ':');
+            if (!end_user) {
+                fputs("Bad command from dom0: no colon\n", stderr);
+                abort();
+            }
+            *end_user++ = '\0';
+        } else {
+            end_user = cmdline;
+        }
+        if (strncmp(end_user, RPC_REQUEST_COMMAND " ", RPC_REQUEST_COMMAND_LEN + 1) != 0) {
+            /* Legacy qrexec behavior: spawn shell directly. */
+            return do_fork_exec(strip_username ? cmdline : NULL, end_user, pid, stdin_fd, stdout_fd, stderr_fd);
+        } else {
+            realcmd = end_user + RPC_REQUEST_COMMAND_LEN + 1;
+            service_descriptor = skip_nogui(realcmd);
+        }
+    }
+    remote_domain = strchr(service_descriptor, ' ');
```

Comment:
If running in dom0 there are more arguments after the domain (like `keywords adminvm`). (If they are intentionally part of remote_domain string please document this)